### ISSUE TITLE
Event Sauce as framework for Third-Party Integrations

### ISSUE THREAD

**Issue by objecttothis** [opensourcepos/opensourcepos#3269]
### Background information

__IMPORTANT: If you choose to ignore this issue report template, your issue will be closed as we cannot help without the requested information.__

Please make sure you tick (add an x between the square brackets with no spaces) the following check boxes:

- [x] Reporting an issue of an unmodified OSPOS installation
- [x] Checked [open and closed issues](https://github.com/opensourcepos/opensourcepos/issues?utf8=%E2%9C%93&q=is%3Aissue) and no similar issue was already reported (please make sure you searched!)
- [x] Read [README](https://github.com/opensourcepos/opensourcepos/blob/master/README.md), [WHATS_NEW](https://github.com/opensourcepos/opensourcepos/blob/master/WHATS_NEW.txt), [INSTALL.md](https://github.com/opensourcepos/opensourcepos/blob/master/INSTALL.md) and [UPGRADE](https://github.com/opensourcepos/opensourcepos/blob/master/UPGRADE.txt)
- [x] Read the [FAQ](https://github.com/opensourcepos/opensourcepos#faq) for any known install and/or upgrade gotchas (in specific PHP extensions installed)
- [x] Read the [wiki](https://github.com/opensourcepos/opensourcepos/wiki)
- [x] Executed any database upgrade scripts if an upgrade pre 3.0.0 (e.g. database/2.4_to_3.0.sql)
- [x] Aware the installation code is in [bintray](https://bintray.com/jekkos/opensourcepos/opensourcepos/view/files?sort=updated&order=asc#files) (see README), and [GitHub master](https://github.com/opensourcepos/opensourcepos/tree/master) is for [developers only](https://github.com/opensourcepos/opensourcepos/wiki/Development-setup) and therefore not complete nor stable

### New Feature

@jekkos had expressed an interest in implementing Event Sauce (https://eventsauce.io/) to provide framework for 3rd party integrations.  This issue will act as a venue for communication about the implementation.

closes #2619 #2608

**Comment by objecttothis** [comment]
@jekkos the latest versions of Event Sauce (1.0.1 and 1.0.0) won't install because they require PHP 8.0.  Event Sauce 0.8.2 does install on PHP 7.4.x.  Should we need to use newer versions of Event Sauce we would need to modify the minimum system requirements of OSPOS to PHP 8.0.x

**Comment by objecttothis** [comment]
@jekkos given that I've implemented event sourcing and you wanted something more robust and that I do not have a lot of experience with the software design part of event sourcing, I'm going to need your help with how you want this designed.  I don't have the time to come up with something and have to redo it several times, so can we come up with a framework as to how you want EventSauce implemented in this project so that I can implement it that way?

I've been reading the EventSauce documentation and the overall concepts are beginning to make sense to me, but it isn't clear to me how you are going to want the various classes structured in the project.

**Comment by objecttothis** [comment]
Here is a conversation I had on slack with the owner of EventSauce

Frank de Jonge  1:06 PM
Hi @object to this, I do not know of any CodeIgniter examples, unfortunately. The files should be placed in a namespace that is relevant to the domain that is being event sourced. I'm working for a PSP, so for me an aggregate that models the payment processing lifecycle would be placed in ProcessingPayments , for new customer onboarding it would be OnboardingNewCustomers and so on.

object to this  1:11 PM
@Frank de Jonge Thank you for your helpful suggestion.  This open source project is a Point of Sale and Inventory management application. I want to make sure I understand your example properly.  Does this mean that I would have more than one Aggregate Root?  For example one for Items and say another for People (customers, suppliers, employees) and then each Aggregate Root would handle only events related to Items or only People?

Frank de Jonge  1:15 PM
seems like you have a domain which is a classic example for bounded contexts, which have their own aggregates per context
1:15
usually Sales and InventoryManagement are separate context
1:17
I would probably not make one "People" oriented aggregate, it's about modelling the lifecycle or process around the things. I'd find it hard to believe that customers suppliers and employees have a lot in common. Those would all probably be individual aggregates
1:18
Not every one of them is probably interesting enough to be an eventsourced aggregate even
1:19
I would also not call something Item, because it's very anonymous. I'd try to find something more specific, Product would be better depending on the context
1:20
That said, this is my initial contextless response

object to this  1:24 PM
Thanks again for your help.  It is appreciated.  I began contributing to the product well after naming conventions were selected, so unfortunately for now it's an inherited problem :wink:.  The main driver for me to implement this is that I have a RESTFUL API that I'm interacting with on another project and I want to take eventsauce events and kick off these API calls.  For now, I'm only interested in items (Products), so any time a Product is modified (price, attribute about the product, etc) or sold (effectively quantity is modified), I want eventsauce to do it's thing which results in an appropriate API call. (edited) 

Frank de Jonge  1:26 PM
Not sure I fully understand. Do you want to effectively use it to model the events and produce them to a queue to do API calls in a consumer?

object to this  1:28 PM
In part, yes.  However others will use the same recorded event to do some other "work".  The idea is to have a repository of events so that whatever integration is using that rather than injecting it's hooks throughout the codebase.

Frank de Jonge  1:31 PM
events are a great integration mechanism

object to this  1:32 PM
Agreed, which is why I'm excited, but as you wrote in your documentation, it's a new paradigm for me.

Frank de Jonge  1:32 PM
I'd recommend researching transactional outboxes, we provide a standard solution for this for doctrine and laravel, which you can use a reference implementation

object to this  1:32 PM
Thank you

Frank de Jonge  1:33 PM
the TL;DR is: You want to only dispatch the events when and if your current database model is persisted. To achieve this, you wrap the model persistence and dispatching into a transaction and use the same DB to dispatch the events
1:33
this makes the dispatching consistent with the model's persistence
1:34
you then consume that DB event buffer with a relay that injects them into your queueing mechanism of choice
1:34
this can be done with PHP, but also using specialized tools such as Debezium (if you use Kafka) (edited)

**Comment by objecttothis** [comment]
@jekkos @daN4cat I just ran across this statement as I was working on the CI 4 upgrade

> Events
>     Hooks have been replaced by Events.
>     Instead of CI3’s $hook['post_controller_constructor'] you now use Events::on('post_controller_constructor', ['MyClass', 'MyFunction']);, with the namespace CodeIgniter\Events\Events;.
>     Events are always enabled, and are available globally.

So, worst case scenario, we need to come up with a different naming convention for whatever event sourcing we come up with.

BUT

Best case scenario, CI4's events functionality fits the model for what we are wanting to do.  I still need to read up on it, but maybe we don't need to implement an event sourcing library after all.  I don't think I was expecting the CI 4 upgrade to resolve both issues.  I was just looking for something to do while @jekkos was busy with other things.

Let me know your thoughts.

**Comment by objecttothis** [comment]
https://codeigniter.com/user_guide/extending/events.html?highlight=events

**Comment by odiea** [comment]
My thoughts are Thank you for doing all of this. Can't wait to get it working.

**Comment by ekardian** [comment]
no problem i can wait centuries, im eterno XD

**Comment by daN4cat** [comment]
@objecttothis just quickly reading CI4 description seems to do the type of event you wanted.
The only question is whether you can have events from all the trigger points you wanted and with the context data you needed.

**Comment by daN4cat** [comment]
Skimmed through the whole post and yes, you can create your own event points.
It seems to me that it's better if we use CI4 support and we add custom events from our controllers.

**Comment by objecttothis** [comment]
> 
> 
> @objecttothis just quickly reading CI4 description seems to do the type of event you wanted.
> The only question is whether you can have events from all the trigger points you wanted and with the context data you needed.

Yeah, this was the reason we didn't use hooks before was because it didn't seem to give us the full data picture that we needed.  I agree that if CI4 built in events works for our implementation, then we should use that instead of adding another point of failure.

The questions that come to mind with CI4 Events are:
- Can an event trigger multiple 3rd party integrations to run?  For example, when a sale is completed, there will likely be a single Event Trigger that will kick off multiple integrations to run.
- Will the integrations have access to the appropriate data when an event triggers?  It doesn't do much good to use CI4 events for event sourcing if the event triggers these integrations but the integrations have no idea what data was changed to do something with it.  For example, if a user edits an item, I will need an event trigger so that I can generate an API call with that updated item, but if the Event doesn't tell me what item, it won't work.  Even better is if the event told me exactly what was changed about the item, but that's not as important.

**Comment by daN4cat** [comment]
Yes and yes.

Pub/Sub has one emitter and multiple listeners.

You can pass params:

```
You can pass any number of arguments to the subscribers by adding them as additional parameters. Subscribers will be given the arguments in the same order as defined:
```

**Comment by jekkos** [comment]
@objecttothis @daN4cat I had a look last week and indeed also saw that CI4 has support for event dispatching. The work @objecttothis seems to align with this and so I agree that it might be better to go with that instead of this 3rd party lib.
